### PR TITLE
Sub process cancellation

### LINKED ISSUE #1119: I can cancel a workflow instance in an embedded subprocess
* for clarity, this issue is separated from #1118 
* this will change a bit how cancellation works; in particular, a response is received as soon as the decision has been made that the workflow instance will be cancelled; actual cancellation will then be performed asynchronously

### PR DESCRIPTION
**Pull request by ThorbenLindhauer**
### Lifecycle

* There is a now a common lifecycle for every workflow element (including the process itself) that has a lifecycle of more than one event. See the `ELEMENT_*` constants in `WorkflowInstanceIntent`.
* The plan is to also have a common state name for single-event elements such as sequence flows and none events in the future.
* There are currently some extra states for the workflow instance (`CREATED`, `CANCELING`) and some have been removed (`CREATED`,`COMPLETED`). To make this consistent (in one way or the other) is task of #1187.

### Index

* There is a single data structure for indexing of workflow instances (see `ElementInstanceIndex`). Every element with a lifecycle is supposed to always be represented there with its latest state.
* In the current implementation, it uses (pretty hacky) Java serialization. We will use RocksDB in the not-so-far future, so I considered the hacks ok => #1188
* The class `ElementInstanceWriter` makes the state updates transparent for the individual step handlers, so state manipulation is in one place.

### Concurrent Processing

* In order to properly handle concurrent commands in a workflow instance (e.g. concurrent instance cancellation and job completion), we check before each event if it should still be handled. E.g. an `ELEMENT_ACTIVATED` record should not be handled if `ELEMENT_TERMINATING` has been written concurrently. This logic is implemented via the guards in `BpmnStepProcessor`.

closes #1119

### CODE REVIEW

**Comment by menski** [comment]
@saig0  pointers: `ElementInstanceIndex`, `ElementInstanceWriter`, `WorkflowInstanceIntent`, `WorkflowInstanceStreamProcessor`

**Comment by saig0** [comment]
Looks good from my side. 

One point I'm not sure about is that the workflow instance and element / activity intents / states are mixed together. I think it could make it hard to decide if this event belongs to a workflow or activity instance. For example, when I implement an export, I want to record the start and end of a workflow instance.

File: broker-core/src/main/java/io/zeebe/broker/workflow/processor/BpmnStepContext.java [hunk 1]
```diff
@@ -71,10 +148,15 @@ public void raiseIncident(ErrorType errorType, String errorMessage) {
         .setErrorMessage(errorMessage);
 
     if (!record.getMetadata().hasIncidentKey()) {
-      streamWriter.writeNewCommand(IncidentIntent.CREATE, incidentCommand);
+      commandWriter.writeNewCommand(IncidentIntent.CREATE, incidentCommand);
     } else {
-      streamWriter.writeFollowUpEvent(
-          record.getMetadata().getIncidentKey(), IncidentIntent.RESOLVE_FAILED, incidentCommand);
+      // TODO: casting is ok for the moment; the problem is rather that we
```
Comment (menski) [inline broker-core/src/main/java/io/zeebe/broker/workflow/processor/BpmnStepContext.java]: @ThorbenLindhauer do we have a issue for this, sounds like something we should work on?

**Review by menski — COMMENTED** [review]
I agree with @saig0 remarks that I think it would be helpful to have a unique event which marks the end of a workflow instance. We still have that on the creating of a workflow instance, i.e. `CREATE`, `CREATED`. But at the end of the instance we have only `ELEMENTE_COMPLETED` or `ELEMENT_TERMINATED`. Which I think is fine but it might be useful to have a unique event at the end which does not require to parse the activity id to distinguish if the workflow instance was completed. 

@ThorbenLindhauer I guess that is what you want to address in #1187?

**Comment by menski** [comment]
We decided to continue the discussion about the events of the workflow instance life cycle in #1187.

**Review by menski — APPROVED** [review]

File: broker-core/src/main/java/io/zeebe/broker/workflow/processor/BpmnStepContext.java (hunk 1 above)
Comment (ThorbenLindhauer) [reply broker-core/src/main/java/io/zeebe/broker/workflow/processor/BpmnStepContext.java]: I added a reference to https://github.com/zeebe-io/zeebe/issues/1033

**Comment by menski** [comment]
bors retry

**Pull request merged**